### PR TITLE
feat: multi-page hearing booking (FPLA-2158)

### PR DESCRIPTION
**Pull request by joez-f**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-2158

### Change description ###

New hearing booking event that is split over multiple pages instead of one page with a collection

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: ccd-definition/CaseEvent/CareSupervision/MultiState-nonprod.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[
```
Comment (lewisbirks) [inline ccd-definition/CaseEvent/CareSupervision/MultiState-nonprod.json:1]: I would add a comment to this and the one being deprecated noting that they should be moved when the toggled on. I use something like
```
"Comment": "To merge/delete when toggling on FPLA-1592"
```

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            caseDetails.getData().put(FIRST_HEARING_FLAG, "Yes");
+        } else {
+            caseDetails.getData().putAll(manageHearingsService.populateInitialFields(caseData));
+        }
+
+        return respond(caseDetails);
+    }
+
+    @PostMapping("/edit-hearing/mid-event")
+    public AboutToStartOrSubmitCallbackResponse populateExistingDraftHearing(
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:75]: If you want to shorten these lines so that the method declaration fits on one you can change the return type to just `CallbackResponse` which `AboutToStartOrSubmitCallbackResponse` implements

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        if ((caseData.getUseExistingHearing() != null) && EDIT_DRAFT.equals(caseData.getUseExistingHearing())) {
+            UUID editedHearingId = getDynamicListValueCode(caseData.getHearingDateList(), mapper);
+
+            caseDetails.getData().put(SELECTED_HEARING_IDS, List.of(editedHearingId));
+
+            hearingBookingElements = manageHearingsService.updateEditedHearingEntry(
+                hearingBooking, editedHearingId, caseData.getHearingDetails());
+        } else {
+            hearingBookingElements = manageHearingsService.appendHearingBooking(
+                defaultIfNull(caseData.getHearingDetails(), List.of()), hearingBooking);
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:137]: I feel like the default getter in case data should return an empty list if it is null

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        HearingBooking hearingBooking = manageHearingsService.buildHearingBooking(caseData);
+
+        if (caseData.getSendNoticeOfHearing() != null && caseData.getSendNoticeOfHearing().equals("Yes")) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:121]: Thoughts on extracting this to a function in case data or the service?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            manageHearingsService.addNoticeOfHearing(caseData, hearingBooking);
+        }
+
+        List<Element<HearingBooking>> hearingBookingElements;
+
+        // Editing previous hearing
+        if ((caseData.getUseExistingHearing() != null) && EDIT_DRAFT.equals(caseData.getUseExistingHearing())) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:128]:
```suggestion
        if (EDIT_DRAFT == caseData.getUseExistingHearing()) {
```

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:128]: Same change to the equals can be applied to the condition in `populateExistingDraftHearing`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingOptions.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
+public enum HearingOptions {
+    NEW_HEARING("Add a new hearing"),
+    EDIT_DRAFT("Edit or issue a draft hearing");
+
+    private final String label;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingOptions.java:12]: Not sure if the label really has a use

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        if (currentHearingBookings.isEmpty()) {
+            return List.of(hearingBookingElement);
+        }
+
+        currentHearingBookings.add(hearingBookingElement);
+        return currentHearingBookings;
+    }
+
+    public List<String> caseFieldsToBeRemoved() {
+        List<String> caseFields = new ArrayList<>();
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:158]: List.of or Set.of can be used instead

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/NoticeOfHearingGenerationService.java [hunk 8]
```diff
@@ -51,4 +54,12 @@ private String getHearingType(HearingBooking hearingBooking) {
             hearingBooking.getTypeDetails();
     }
 
+    //feature toggle
+    private String getHearingNotes(String multiHearingNotes, String singleHearingNotes) {
+        if (featureToggleService.isMultiPageHearingEnabled()) {
+            return defaultIfNull(multiHearingNotes, "");
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/NoticeOfHearingGenerationService.java:60]: The isBlank in the document template should be able to handle nulls

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                        .value(hearingBooking)
+                        .build();
+                }
+                return hearingBookingElement;
+            }).collect(Collectors.toList());
+    }
+
+    public List<Element<HearingBooking>> appendHearingBooking(List<Element<HearingBooking>> currentHearingBookings,
+                                                              HearingBooking hearingBooking) {
+        Element<HearingBooking> hearingBookingElement = Element.<HearingBooking>builder()
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:144]: ElementUtils.element(...)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        hearingBooking.setNoticeOfHearing(DocumentReference.buildFromDocument(document));
+    }
+
+    public List<Element<HearingBooking>> updateEditedHearingEntry(HearingBooking hearingBooking,
+                                                                  UUID hearingId,
+                                                                  List<Element<HearingBooking>> hearings) {
+        return hearings.stream()
+            .map(hearingBookingElement -> {
+                if (hearingBookingElement.getId().equals(hearingId)) {
+                    hearingBookingElement = Element.<HearingBooking>builder()
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:133]: ElementUtils

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+        if (hearingBookingElement.isPresent()) {
+            return hearingBookingElement.get().getValue();
+        }
+
+        return HearingBooking.builder().build();
+    }
+
+    public Map<String, Object> populateHearingCaseFields(HearingBooking hearingBooking) {
+        Map<String, Object> data = new HashMap<>();
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:96]: Map.of can be used here

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    private final HearingVenueLookUpService hearingVenueLookUpService;
+    private final HearingBookingService hearingBookingService;
+
+    public static final String FIRST_HEARING_FLAG = "firstHearingFlag";
+    public static final String HEARING_DATE_LIST = "hearingDateList";
+
+    public Map<String, Object> populateInitialFields(CaseData caseData) {
+        Map<String, Object> data = new HashMap<>();
+        List<Element<HearingBooking>> futureHearings = hearingBookingService.getFutureHearings(
+            defaultIfNull(caseData.getHearingDetails(), List.of()));
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:53]: again altering the getter in case data would make this more simple

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java [hunk 13]
```diff
@@ -64,6 +65,10 @@
         return hearingDetails.stream().filter(this::isPastHearing).collect(toList());
     }
 
+    public List<Element<HearingBooking>> getFutureHearings(List<Element<HearingBooking>> hearingDetails) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java:68]: Keep in the service or can be in case data?

**Review by lewisbirks — COMMENTED** [review]
Had a quick look over

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java (hunk 13 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java:68]: Case data 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java (hunk 3 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:137]: Will handle that in https://tools.hmcts.net/jira/browse/FPLA-2280 during refactor + cleanup

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingOptions.java (hunk 6 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingOptions.java:12]: agreed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/NoticeOfHearingGenerationService.java (hunk 8 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/NoticeOfHearingGenerationService.java:60]: will check that

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java (hunk 12 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:53]: or just moving the whole thing to caseData like you suggested 😄

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java (hunk 4 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java:121]: what about
        if ("Yes".equals(caseData.getSendNoticeOfHearing()))
?

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ManageHearingsController.java]: or that, makes much more sense 🙂

File: codecept.conf.js [hunk 14]
```diff
@@ -65,6 +65,7 @@ exports.config = {
     notifyGatekeeperEventPage: './e2e/pages/events/sendCaseToGatekeeperEvent.page.js',
     createNoticeOfProceedingsEventPage: './e2e/pages/events/createNoticeOfProceedingsEvent.page.js',
     addHearingBookingDetailsEventPage: './e2e/pages/events/addHearingBookingDetailsEvent.page.js',
```
Comment (lewisbirks) [inline codecept.conf.js:67]: Going to remove this?

File: e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js [hunk 15]
```diff
@@ -60,7 +61,35 @@ Scenario('Gatekeeper enters allocation decision', async (I, caseViewPage, enterA
...
+  await manageHearingsEventPage.enterVenue(hearingDetails[0]);
+  await I.retryUntilExists(() => I.click('Continue'), '#judgeAndLegalAdvisor_judgeTitle');
+  await manageHearingsEventPage.enterJudgeAndLegalAdvisorDetails(hearingDetails[0]);
+  await I.retryUntilExists(() => I.click('Continue'), '#sendNoticeOfHearing');
+  await manageHearingsEventPage.sendNoticeOfHearingWithNotes(hearingDetails[0].additionalNotes);
+  await I.completeEvent('Save and continue');
+  I.seeEventSubmissionConfirmation(config.administrationActions.manageHearings);
+  caseViewPage.selectTab(caseViewPage.tabs.hearings);
+
+  I.wait(10);
```
Comment (lewisbirks) [inline e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js:76]: Still required in only one pass? We could have this flow select the no option for sending notice of proceedings then you wouldn't have the issue.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingOptions.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.reform.fpl.enums;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingOptions.java:7]: These annotations are not required

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+            && caseData.getPreviousHearingVenue().getUsePreviousVenue().equals("Yes")) {
+
+            PreviousHearingVenue previousVenueForEditedHearing = caseData.getPreviousHearingVenue();
+
+            caseData.setPreviousVenueId(hearingVenueLookUpService.getVenueId(
+                previousVenueForEditedHearing.getPreviousVenue()));
+        }
+    }
+
+    public HearingBooking buildHearingBooking(CaseData caseData) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ManageHearingsService.java:125]: If you do the judge and legal advisor migration stuff in these private methods then you wouldn't need to get the selected judge when building the notice of hearing I think. And it would mean that the tabs look more consistent with their current design.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/JudgeAndLegalAdvisor.java [hunk 18]
```diff
@@ -16,7 +16,7 @@
     private final String otherTitle;
     private final String judgeLastName;
     private final String judgeFullName;
-    private final String legalAdvisorName;
+    private String legalAdvisorName;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/JudgeAndLegalAdvisor.java:19]: unnecessary change?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java [hunk 19]
```diff
@@ -34,7 +34,8 @@
 public class HearingBooking {
     private final HearingType type;
     private final String typeDetails;
-    private final String venue;
+    private String venue;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/HearingBooking.java:37]: unnecessary change?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 20]
```diff
@@ -507,4 +519,26 @@ public DynamicList buildDynamicHearingList() {
     public DynamicList buildDynamicHearingList(UUID selected) {
         return asDynamicList(getHearingDetails(), selected, hearingBooking -> hearingBooking.toLabel(DATE));
     }
+
+    private final HearingType hearingType;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java:523]: Maybe not now but I noticed that there was a tech debt for clean up. Thoughts on combining all these new fields into one model class and then annotating that class in here with `@JsonUnwrapped`. This means that then all the fields are cleaned up in one class from our perspective, and from the perspective of CCD they are still all top level data fields.
e.g.
```java
public class ManageHearingBuilder {
    private final HearingType hearingType;
    private final String hearingTypeDetails;
    private final String hearingVenue;
    private final Address hearingVenueCustom;
    private final String firstHearingFlag; //also used for logic surrounding legacy hearings
    private final PreviousHearingVenue previousHearingVenue;
    private String previousVenueId;
    private final String noticeOfHearingNotes;
    private final Object hearingDateList;
    private final String hasExistingHearings;

    @TimeNotMidnight(message = "Enter a valid start time", groups = HearingDatesGroup.class)
    @Future(message = "Enter a start date in the future", groups = HearingDatesGroup.class)
    private final LocalDateTime hearingStartDate;

    @TimeNotMidnight(message = "Enter a valid end time", groups = HearingDatesGroup.class)
    @Future(message = "Enter an end date in the future", groups = HearingDatesGroup.class)
    private final LocalDateTime hearingEndDate;
    private final String sendNoticeOfHearing;
    private final HearingOptions hearingOption;
    private final List<ProceedingType> proceedingType;
}
```
then in case data it would be something like
```java
    private ManageHearingBuilder manageHearingBuilder;
```

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java:523]: This object could also contain the set of keys to be removed as a static field.

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java:523]: This could also replace the use of a map to populate initial fields, then in the controller just convert this object to a map and add that map to data.
```java
ManageHearingBuilder builder = populateHearingCaseFields(hearingBooking);
caseDetails.getData().putAll(mapper.convertValue(builder, new TypeReference<>(){});

File: codecept.conf.js (hunk 14 above)
Comment (lewisbirks) [reply codecept.conf.js:67]: Just seen that there is a ticket to clean stuff up, this going to be part of it?

File: e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js (hunk 15 above)
Comment (joez-f) [reply e2e/tests/gatekeeperAdministersCaseAfterGatekeeping_test.js:76]: Issue arises because first hearing triggers populate SDO dates, which affects case data, which will cause case data modified outside transaction error

File: e2e/config.js [hunk 21]
```diff
@@ -111,6 +111,7 @@ module.exports = {
     amendAttendingHearing: 'Attending the hearing',
     amendRepresentatives: 'Manage representatives',
     addHearingBookingDetails: 'Add hearing details',
```
[the remark this answers is not in the document]
Comment (joez-f) [reply e2e/config.js:113]: In FPLA-2280 will cleanup

File: codecept.conf.js (hunk 14 above)
Comment (joez-f) [reply codecept.conf.js:67]: In FPLA-2280 will cleanup

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/JudgeAndLegalAdvisor.java (hunk 18 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/JudgeAndLegalAdvisor.java]: yep 🙊

**Review by lewisbirks — APPROVED** [review]

**Review by nraghu32 — APPROVED** [review]
